### DISCUSSION TITLE
MaxAttempts not being used in InsertOpts

### DISCUSSION THREAD

**Discussion by Zibbp** [riverqueue/river#412] [Q&A] [answered]
**Version: v0.8.0**
Hi, I'm trying to get the `MaxAttempts` option to take effect when being set on the job's arguments like so.
```go
type WatchdogArgs struct{}

func (WatchdogArgs) Kind() string { return "archive-watchdog" }

func (w *WatchdogArgs) InsertOpts() river.InsertOpts {
	return river.InsertOpts{
		MaxAttempts: 1,
		Queue:       "default",
	}
}

func (w *WatchdogArgs) Timeout(job *river.Job[WatchdogArgs]) time.Duration {
	return 45 * time.Second
}

type WatchdogWorker struct {
	river.WorkerDefaults[WatchdogArgs]
}

func (w *WatchdogWorker) Work(ctx context.Context, job *river.Job[WatchdogArgs]) error {
...
}
```

This job is setup has a periodic task 
```go
	periodicJobs := []*river.PeriodicJob{
		river.NewPeriodicJob(
			river.PeriodicInterval(1*time.Minute),
			func() (river.JobArgs, *river.InsertOpts) {
				return WatchdogArgs{}, nil
			},
			&river.PeriodicJobOpts{RunOnStart: true},
		),
	}

...


	riverClient, err := river.NewClient(rc.RiverPgxDriver, &river.Config{
		Workers:              workers,
		PeriodicJobs:         periodicJobs,
	})
```

When the job is inserted by the client every minute, it has `max_attempts` set to 25.
![image](https://github.com/riverqueue/river/assets/21207065/e4d17012-8175-4c1e-a6b1-0d4ddceb1ce6)

Is there something I'm missing? The `InsertOpts` on the job should be overriding the default.

**Comment by brandur** [comment, marked as answer]
Can you try giving `InsertOpts` a non-pointer receiver and see if that fixes it?

``` go
func (w WatchdogArgs) InsertOpts() river.InsertOpts {
	return river.InsertOpts{
		MaxAttempts: 1,
		Queue:       "default",
	}
}
```

(See the `(w WatchdogArgs)` part where I removed the pointer.)

It may be that since a non-pointer implements the `JobArgs` interface (you have `func (WatchdogArgs) Kind()`), then you also need a non-pointer to implement `JobArgsWithInsertOpts` which is what extracts an `InsertOpts`.

**Reply by Zibbp** [reply]
That was it, thank you.
